src/actions/SupplierCreditActions.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * mSupply Mobile
+ * Sustainable Solutions (NZ) Ltd. 2020
+ */
+
+import { batch } from 'react-redux';
+import { createRecord, UIDatabase } from '../database';
+import { refreshData } from '../pages/dataTableUtilities/actions/tableActions';
+import { ROUTES } from '../navigation/constants';
+
+export const SUPPLIER_CREDIT_ACTIONS = {
+  CREATE_FROM_ITEM: 'SupplierCredit/createFromItem',
+  CLOSE: 'SupplierCredit/close',
+  SORT: 'SupplierCredit/sort',
+  EDIT_RETURN_AMOUNT: 'SupplierCredit/editReturnAmount',
+};
+
+const sort = sortKey => ({ type: SUPPLIER_CREDIT_ACTIONS.SORT, payload: { sortKey } });
+
+const close = () => ({ type: SUPPLIER_CREDIT_ACTIONS.CLOSE });
+
+const create = () => (dispatch, getState) => {
+  const { supplierCredit, user } = getState();
+  const { batches } = supplierCredit;
+  const { currentUser } = user;
+
+  // Only work with the batches whose return amount is greater than 0.
+  const batchesToReturn = batches.reduce(
+    (acc, itemBatch) => (itemBatch.returnAmount ? [...acc, itemBatch] : acc),
+    []
+  );
+
+  // Group the batches with a return amount by supplier to make a credit for each
+  // supplier grouping { supplierId1: [batch1, batch2, ... batchn], supplierId2: [...], ...}
+  const batchesGroupedBySupplier = batchesToReturn.reduce((groupings, itemBatch) => {
+    const { id: suppliersId } = itemBatch?.itemBatch.supplier;
+    const suppliersGroup = groupings[suppliersId];
+
+    if (suppliersGroup) return { ...groupings, [suppliersId]: suppliersGroup.push(itemBatch) };
+    return { ...groupings, [suppliersId]: [itemBatch] };
+  }, {});
+
+  UIDatabase.write(() => {
+    Object.entries(batchesGroupedBySupplier).forEach(([supplierId, suppliersBatches]) => {
+      const returnSum = -suppliersBatches.reduce(
+        (total, { costPrice, returnAmount }) => total + costPrice * returnAmount,
+        0
```

Comment:
Nice!

Nit: can make it a bit nicer by shifting the negative into the reducer:

```
const returnSum = suppliersBatches.reduce(
        (total, { costPrice, returnAmount }) => total - (costPrice * returnAmount),
        0
);
```

Reply:
c7cd030ecb44c1842b6d7edee6bf060aa9d3c11b